### PR TITLE
Filter out unsupported proposals

### PR DESCRIPTION
**Pull request by tadovas**
The main idea how to handle proposals of uknown service,payment or contact types.
Proposal is considered unsupported if one of the following is true:

- Unknown service type
- Unnown payment type
- **All** contact types in contact list are unknown.

Such proposals will be filtered out and not returned by tequilapi

### CODE REVIEW

File: service_discovery/dto/contact.go [hunk 1]
```diff
@@ -30,9 +30,68 @@ func (list ContactList) MarshalJSON() ([]byte, error) {
...
+	return
+}
+
+func unserializeContact(contactType string, rawMessage *json.RawMessage) ContactDefinition {
+	fn, ok := contactDefinitionMap[contactType]
+	if !ok {
+		return UnsupportedContactType{}
+	}
+	definition, er := fn(rawMessage)
+	if er != nil {
```
Comment (zolia) [inline service_discovery/dto/contact.go]: er? :)

Comment (tadovas) [reply service_discovery/dto/contact.go]: really? :) fine :)

File: service_discovery/dto/payment_method.go [hunk 2]
```diff
@@ -17,9 +17,49 @@
...
 package dto
 
-import "github.com/mysteriumnetwork/node/money"
+import (
+	"encoding/json"
 
+	"github.com/mysteriumnetwork/node/money"
+)
+
+// PaymentMethod is a marker interface for all types of payment methods
```
Comment (zolia) [inline service_discovery/dto/payment_method.go]: just interface?

File: service_discovery/dto/payment_method.go [hunk 3]
```diff
@@ -17,9 +17,49 @@
...
+var _ PaymentMethod = UnsupportedPaymentMethod{}
+
+// PaymentMethodUnserializer is function type which takes raw json message and returns deserialized payment method
+type PaymentMethodUnserializer func(*json.RawMessage) (PaymentMethod, error)
+
+// service payment method unserializer registry
+var paymentMethodMap = make(map[string]PaymentMethodUnserializer, 0)
+
+// RegisterPaymentMethodUnserializer registers unserializer for specified payment method type
+func RegisterPaymentMethodUnserializer(paymentMethod string, unserializer func(*json.RawMessage) (PaymentMethod, error)) {
```
Comment (zolia) [inline service_discovery/dto/payment_method.go]: Should it return error? Don't see its usage.

File: service_discovery/dto/service_definition.go [hunk 4]
```diff
@@ -17,6 +17,43 @@
 
 package dto
 
+import "encoding/json"
+
+// ServiceDefinition interface is marker interface for all service definition types
```
Comment (zolia) [inline service_discovery/dto/service_definition.go]: not marker here too.

File: service_discovery/dto/payment_method.go (hunk 3 above)
Comment (tadovas) [reply service_discovery/dto/payment_method.go]: That's the unserializer function which returns error. Register function doesn't return anything

File: service_discovery/dto/payment_method.go (hunk 2 above)
Comment (tadovas) [reply service_discovery/dto/payment_method.go]: Removed markers

File: service_discovery/dto/service_proposal.go [hunk 5]
```diff
@@ -28,6 +27,8 @@ const (
 	proposalFormat = "service-proposal/v1"
 )
 
+// ServiceProposal is top level structure which is presented to marked place by service provider, and looked up by service consumer
```
Comment (zolia) [inline service_discovery/dto/service_proposal.go]:
```suggestion
// ServiceProposal is top level structure which is presented to marketplace by service provider, and looked up by service consumer
```

File: service_discovery/dto/payment_method.go [hunk 6]
```diff
@@ -17,9 +17,49 @@
...
+	"github.com/mysteriumnetwork/node/money"
+)
+
+// PaymentMethod is a interface for all types of payment methods
 type PaymentMethod interface {
 	// Service price per unit of metering
 	GetPrice() money.Money
 }
+
+// UnsupportedPaymentMethod represents payment method which is uknown to node (i.e. not registered)
```
Comment (soffokl) [inline service_discovery/dto/payment_method.go]:
```suggestion
// UnsupportedPaymentMethod represents payment method which is unknown to node (i.e. not registered)
```

File: service_discovery/dto/payment_method.go [hunk 7]
```diff
@@ -17,9 +17,49 @@
...
+	panic("not supported")
+}
+
+var _ PaymentMethod = UnsupportedPaymentMethod{}
+
+// PaymentMethodUnserializer is function type which takes raw json message and returns deserialized payment method
+type PaymentMethodUnserializer func(*json.RawMessage) (PaymentMethod, error)
+
+// service payment method unserializer registry
+var paymentMethodMap = make(map[string]PaymentMethodUnserializer, 0)
```
Comment (soffokl) [inline service_discovery/dto/payment_method.go]: Can we add TODO here to replace this global map too?

File: service_discovery/dto/service_definition.go [hunk 8]
```diff
@@ -17,6 +17,43 @@
...
+	panic("not supported")
+}
+
+var _ ServiceDefinition = UnsupportedServiceDefinition{}
+
+// ServiceDefinitionUnserializer defines function to register for concrete service definition
+type ServiceDefinitionUnserializer func(*json.RawMessage) (ServiceDefinition, error)
+
+// service definition unserializer registry
+var serviceDefinitionMap = make(map[string]ServiceDefinitionUnserializer, 10)
```
Comment (soffokl) [inline service_discovery/dto/service_definition.go]: Why do we need to preallocate map with a size of 10 here?

File: service_discovery/dto/service_definition.go [hunk 9]
```diff
@@ -17,6 +17,43 @@
...
+	//no location available - should never be called
+	panic("not supported")
+}
+
+var _ ServiceDefinition = UnsupportedServiceDefinition{}
+
+// ServiceDefinitionUnserializer defines function to register for concrete service definition
+type ServiceDefinitionUnserializer func(*json.RawMessage) (ServiceDefinition, error)
+
+// service definition unserializer registry
```
Comment (soffokl) [inline service_discovery/dto/service_definition.go]: I'd like to have TODO here too, to replace global map in the future.

File: service_discovery/dto/service_proposal.go [hunk 10]
```diff
@@ -171,19 +78,49 @@ func (proposal *ServiceProposal) UnmarshalJSON(data []byte) error {
...
+// SetProviderContact updates service proposal description with general data
+func (proposal *ServiceProposal) SetProviderContact(providerID identity.Identity, providerContact Contact) {
+	proposal.Format = proposalFormat
+	// TODO This will be generated later
+	proposal.ID = 1
+	proposal.ProviderID = providerID.Address
+	proposal.ProviderContacts = ContactList{providerContact}
+}
+
+// IsSupported function returns true if this service proposal can be used for connections by service consumer
```
Comment (soffokl) [inline service_discovery/dto/service_proposal.go]:
```suggestion
// IsSupported returns true if this service proposal can be used for connections by service consumer
```

File: service_discovery/dto/service_definition.go (hunk 8 above)
Comment (tadovas) [reply service_discovery/dto/service_definition.go]: No idea. It's not my code. But it can be safely removed anyway.

**Comment by tadovas** [comment]
@zolia @vkuznecovas @soffokl ready again

**Pull request merged**